### PR TITLE
Calculating prio fees by account

### PR DESCRIPTION
**Pull request by godmodegalactus**

### CODE REVIEW

File: lite-rpc/src/bridge.rs [hunk 1]
```diff
@@ -597,4 +615,41 @@ impl LiteRpcServer for LiteBridge {
 
         Ok(())
     }
+
+    async fn get_latest_account_priofees(
+        &self,
+        account: String,
+        method: Option<PrioritizationFeeCalculationMethod>,
```
Comment (grooviegermanikus) [inline lite-rpc/src/bridge.rs:622]: would not use an enum+optional. couldn't you just define enum with appropriate variants?

File: lite-rpc/src/bridge.rs [hunk 2]
```diff
@@ -597,4 +615,41 @@ impl LiteRpcServer for LiteBridge {
...
+            ))
+        }
+    }
+
+    async fn latest_account_priofees_subscribe(
+        &self,
+        _pending: PendingSubscriptionSink,
+        _account: String,
+    ) -> SubscriptionResult {
+        todo!()
```
Comment (grooviegermanikus) [inline lite-rpc/src/bridge.rs:653]: todo?

File: lite-rpc/src/rpc.rs [hunk 3]
```diff
@@ -237,9 +238,22 @@ pub trait LiteRpc {
...
+    async fn get_latest_block_priofees(
+        &self,
+        method: Option<PrioritizationFeeCalculationMethod>,
+    ) -> crate::rpc::Result<RpcResponse<PrioFeesStats>>;
 
     /// subscribe to prio fees distribution per block; uses confirmation level "confirmed"
     #[subscription(name = "blockPrioritizationFeesSubscribe" => "blockPrioritizationFeesNotification", unsubscribe="blockPrioritizationFeesUnsubscribe", item=PrioFeesStats)]
     async fn latest_block_priofees_subscribe(&self) -> SubscriptionResult;
+
+    #[method(name = "getLatestAccountsPrioFees")]
```
Comment (grooviegermanikus) [inline lite-rpc/src/rpc.rs:250]: nit: shouldn't it be _Account_ and not _Accounts_ ?

File: prioritization_fees/src/block_priofees.rs [hunk 4]
```diff
@@ -29,13 +26,33 @@ pub struct PrioFeesService {
...
+        lock.last_key_value()
+            .map(|(slot, value)| (*slot, value.calculate_stats()))
+    }
+
+    pub async fn get_last_n_priofees_aggregate(&self, nb: usize) -> Option<(Slot, PrioFeesStats)> {
+        let lock = self.block_fees_store.recent.read().await;
+        let last_slot = match lock.last_key_value().map(|x| *x.0) {
+            Some(slot) => slot,
+            None => {
+                return None;
```
Comment (grooviegermanikus) [inline prioritization_fees/src/block_priofees.rs:38]: maybe you should also abort if there are less than `nb` items in the map?

File: prioritization_fees/src/prioritization_fee_calculation_method.rs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
+#[serde(rename_all = "camelCase")]
+pub enum PrioritizationFeeCalculationMethod {
+    #[default]
+    Latest,
+    Last(usize),
```
Comment (grooviegermanikus) [inline prioritization_fees/src/prioritization_fee_calculation_method.rs:8]: maybe we can find a more clear naming .. Latest..Last is not really catchy

File: prioritization_fees/src/prioritization_fee_data.rs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        Self {
+            priority: value.0,
+            cu_consumed: value.1,
+        }
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct BlockPrioData {
+    pub tx_prioritization: Vec<PrioFeesData>,
```
Comment (grooviegermanikus) [inline prioritization_fees/src/prioritization_fee_data.rs:23]: nit: why do you use the term _prioritization_? the list contains the raw prio fees which have impact on the prioritization but that's not the same

File: prioritization_fees/src/account_prio_service.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
+use solana_lite_rpc_core::types::BlockStream;
+use solana_sdk::pubkey::Pubkey;
+use tokio::task::JoinHandle;
+
+use crate::{account_priofees::AccountPrioStore, rpc_data::AccountPrioFeesStats};
+use tokio::sync::broadcast::error::RecvError::{Closed, Lagged};
+
+#[derive(Clone)]
```
Comment (grooviegermanikus) [inline prioritization_fees/src/account_prio_service.rs:8]: cloning service + store feels bit wrong but maybe that's just my preference

File: prioritization_fees/src/account_priofees.rs [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+                match accounts_by_prioritization_write.get_mut(write_lock) {
+                    Some(acc_vec) => {
+                        acc_vec.push(value);
+                    }
+                    None => {
+                        accounts_by_prioritization_write.insert(*write_lock, vec![value]);
+                    }
+                }
+
+                match accounts_by_prioritization_read_write.get_mut(write_lock) {
```
Comment (grooviegermanikus) [inline prioritization_fees/src/account_priofees.rs:66]: c&p: should be _read_lock_

File: prioritization_fees/src/account_priofees.rs [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+                    None => {
+                        accounts_by_prioritization_read_write.insert(*write_lock, vec![value]);
+                    }
+                }
+            }
+
+            for readlock in &transaction.readable_accounts {
+                match accounts_by_prioritization_read_write.get_mut(readlock) {
+                    Some(acc_vec) => {
+                        acc_vec.push(value);
```
Comment (grooviegermanikus) [inline prioritization_fees/src/account_priofees.rs:79]: there will be two items added to vec if `for write_lock` and `for readlock` is matching. is that correct?

File: prioritization_fees/src/account_priofees.rs [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+                    };
+                    prio_fee.stats_by_slot.insert(slot, block_prio_data);
+                    self.account_by_prio_fees_writeonly
+                        .insert(account, prio_fee);
+                }
+            }
+        }
+
+        // cleanup old data
+        let last_slot = produced_block
```
Comment (grooviegermanikus) [inline prioritization_fees/src/account_priofees.rs:141]: `last_slot` is very misleading (even worse: there is `self.last_slot` with completely different semantics from what I understand).

maybe call this one `min_slot_to_retain`

File: prioritization_fees/src/account_priofees.rs [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+            .for_each(cleanup_functor);
+        self.last_slot
+            .store(slot, std::sync::atomic::Ordering::Relaxed);
+
+        self.account_by_prio_fees_writeonly
+            .iter_mut()
+            .for_each(cleanup_functor);
+    }
+
+    pub fn get_latest_stats(&self, account: &Pubkey) -> (u64, AccountPrioFeesStats) {
```
Comment (grooviegermanikus) [inline prioritization_fees/src/account_priofees.rs:163]: u64 -> Slot?

**Review by grooviegermanikus — CHANGES_REQUESTED** [review]
some minor comments

File: prioritization_fees/src/account_prio_service.rs (hunk 7 above)
Comment (godmodegalactus) [reply prioritization_fees/src/account_prio_service.rs:8]: Should be ok, the store is not cloned just a pointer.

File: prioritization_fees/src/account_priofees.rs (hunk 9 above)
Comment (godmodegalactus) [reply prioritization_fees/src/account_priofees.rs:79]: Yeah one is only write_lock. other is both write_lock and read lock.

File: prioritization_fees/src/block_priofees.rs (hunk 4 above)
Comment (godmodegalactus) [reply prioritization_fees/src/block_priofees.rs:38]: will do it later

**Pull request merged**